### PR TITLE
chore: v3 interface

### PR DESCRIPTION
**Pull request by mpetrunic**
## Related Issue Or Context

Closes: #<issue>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [ ] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

**Comment by mpetrunic** [comment]
Think it's a good starting point for discussion. What's missing, whats, confusing etc

Still TODOs:
- add spectre related methods
- add error codes on each method

File: packages/core/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+export type Route = {
+  fromDomain: Domain;
+  toDomain: Domain;
+  resource: Resource;
+  feeHandler: FeeHandler;
+};
+
+export type TransferStatus = "pending" | "executed" | "failed";
+
+export type TransferStatusResponse = {
```
Comment (mpetrun5) [inline packages/core/src/index.ts:88]: Should this also return the details of the transactions (sourceHash, destinationHash...)?

Comment (mpetrunic) [reply packages/core/src/index.ts:88]: I would love to add it but afaik indexer doesn't have that in response?! cc @MakMuftic

**Comment by mpetrun5** [comment]
On the Spectre changes:
 - the only user facing difference is the security model [param](https://github.com/sygmaprotocol/sygma-x-solidity/blob/30cb6c431c34ad265f0f5ad95498be57863dfd11/src/contracts/Router.sol#L97)
 
For that we probably need to add some predefined enum when creating transfer that defines type of transfer verifier (MPC, MPC+Zipline, Zipline+Spectre...). TBD if that definition should we store somewhere or hardcode in SDK.

**Comment by mpetrunic** [comment]
> On the Spectre changes:
> 
> * the only user facing difference is the security model [param](https://github.com/sygmaprotocol/sygma-x-solidity/blob/30cb6c431c34ad265f0f5ad95498be57863dfd11/src/contracts/Router.sol#L97)
> 
> For that we probably need to add some predefined enum when creating transfer that defines type of transfer verifier (MPC, MPC+Zipline, Zipline+Spectre...). TBD if that definition should we should store somewhere or hardcode in SDK.

Does that param affect fee calculation?

**Comment by mpetrun5** [comment]
> > On the Spectre changes:
> > 
> > * the only user facing difference is the security model [param](https://github.com/sygmaprotocol/sygma-x-solidity/blob/30cb6c431c34ad265f0f5ad95498be57863dfd11/src/contracts/Router.sol#L97)
> > 
> > For that we probably need to add some predefined enum when creating transfer that defines type of transfer verifier (MPC, MPC+Zipline, Zipline+Spectre...). TBD if that definition should we should store somewhere or hardcode in SDK.
> 
> Does that param affect fee calculation?

Yes, it does. That is the non user facing change. 
The fee router map was extended so the fee can be setup per security model (https://github.com/sygmaprotocol/sygma-x-solidity/blob/30cb6c431c34ad265f0f5ad95498be57863dfd11/src/contracts/handlers/FeeHandlerRouter.sol#L17) and the feeData now should have that param as well.

File: packages/core/src/index.ts (hunk 1 above)
Comment (mpetrun5) [reply packages/core/src/index.ts:88]: I'm pretty sure we return the whole `Transfer` object from the db so it should all be there already.

Comment (mpetrunic) [reply packages/core/src/index.ts:88]: Maybe somebody can share openapi link for indexer? :)

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  status: TransferStatus;
+  explorerUrl: string;
```
Comment (wainola) [inline packages/core/src/index.ts:90]: nit: maybe `transferDetailUrl`?

File: packages/core/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+export interface SubstrateConfig extends BaseConfig<Network.SUBSTRATE> {
+  handlers: Array<Handler>;
+}
+
+export type Handler = {
+  type: ResourceType;
+  address: string;
+};
+
+export interface RawConfig {
```
Comment (wainola) [inline packages/core/src/index.ts:129]: nit: maybe `SygmaConfig`?

File: packages/evm/src/fungible.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  type: FeeHandlerType;
+  handlerAddress: string;
+  tokenAddress?: string;
+  percentage?: number;
+  minFee?: bigint;
+  maxFee?: bigint;
+};
+
+/**
+ * Return amount of liqudiity tokens on resource handler
```
Comment (wainola) [inline packages/evm/src/fungible.ts:33]:
```suggestion
 * Return amount of liquidity tokens on resource handler
```

File: packages/core/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable @typescript-eslint/require-await */
+
+export enum RouteType {
+  GMP = "gmp",
+  FUNGIBLE = "fungible",
```
Comment (wainola) [inline packages/core/src/index.ts:6]: what about non-fungible?

File: packages/core/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+/**
+ * Returns  supported routes originating from given source domain.
+ * @param source Either caip2 identifier, chainId or sygmaId
+ * @param options Allows selecting bridge instance (mainnet by default) and filtering routes by type.
+ */
+export async function getRoutes(
+  source: string | number | Domain,
+  options?: {
+    env?: Environment;
```
Comment (wainola) [inline packages/core/src/index.ts:155]: do we need env as optional parameter if supported routes per source are bound to the environment? 🤔

File: packages/core/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  LOCAL = "local",
+  DEVNET = "devnet",
+  TESTNET = "testnet",
+  MAINNET = "mainnet",
+}
+
+export enum FeeHandlerType {
+  BASIC = "basic",
+  PERCENTAGE = "percentage",
+  UNDEFINED = "undefined",
```
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:19]: I don't think it should but I think it's theoretically possible if no-one sets one. But that route is not usable in that case I think.

File: packages/core/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+  SUBSTRATE = "substrate",
+}
+
+export type Domain = {
+  sygmaId: number;
+  // https://chainagnostic.org/CAIPs/caip-2
+  caipId: string;
+  chainId: number;
+  name: string;
+  iconUrl?: string;
```
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:33]: What do you mean collide? SDK will use that indexer API to augment chain data

File: packages/core/src/index.ts (hunk 6 above)
Comment (mpetrunic) [reply packages/core/src/index.ts:155]: I think you are right, will revert that

File: packages/core/src/index.ts (hunk 2 above)
Comment (MakMuftic) [reply packages/core/src/index.ts:90]: `explorerUrl` imo.

File: packages/core/src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (wainola) [reply packages/core/src/index.ts:33]: On second thought I got what you mean here, nvm my quesrtion.

**Comment by MakMuftic** [comment]
@mpetrunic I think we should also include property for selected security mechanisms, as we will need this in the future. For now, it can even be hardcoded to the MPC, but we will not break function signatures in the future. wdyt?

File: packages/evm/src/fungible.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ * @param resource
+ */
+export function getLiquidity(
+  provider: Eip1193Provider,
+  resource: EvmResource,
+): Promise<bigint> {
+  throw new Error("Method not implemented");
+}
+
+export function createEvmFungibleAssetTransfer(
```
Comment (saadjhk) [inline packages/evm/src/fungible.ts:44]: is spectre part of the release, if so how would that be specified

Comment (mpetrunic) [reply packages/evm/src/fungible.ts:55]: Added SecurityModel params. But only option is MPC. Once we add support for spectre, it will only change SDK code paths

File: packages/core/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  throw new Error("Not implemented");
+}
+
+/**
+ * Returns  supported routes originating from given source domain.
+ * @param source Either caip2 identifier, chainId or sygmaId
+ * @param options Allows selecting bridge instance (mainnet by default) and filtering routes by type.
+ */
+export async function getRoutes(
+  source: string | number | Domain,
```
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:159]: Apparently, we don't need it since it is infered from source network env

File: packages/evm/src/generic.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+ * @param sourceDomain
+ * @param sourceNetworkProvider
+ * @param destinationDomain
+ * @param destinationContractAddress
+ * @param destinationContractAbi
+ * @param functionName
+ * @param functionParameters
+ * @param gasLimit
+ */
+export function createEvmContractCallTransfer(
```
Comment (saadjhk) [inline packages/evm/src/generic.ts:79]: `createEvmTransferContractCall` how about

File: packages/core/src/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (saadjhk) [reply packages/core/src/index.ts:159]: @mpetrunic in that case `core` shouldn't have a function for fetching, caching and providing configuration w.r.t the environment

File: packages/core/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+  explorerUrl: string;
+  sourceHash: string;
+  destinationHash: string;
+  depositNonce: number;
+  fromDomainId: number;
+  toDomainId: number;
+};
+
+export interface BaseConfig<T> {
+  id: number;
```
Comment (saadjhk) [inline packages/core/src/index.ts:102]: we will need `caipId` here as well

File: packages/evm/src/generic.ts (hunk 11 above)
Comment (mpetrunic) [reply packages/evm/src/generic.ts:79]: Or `createEvmCrossChainContractCall`, either works for me

File: packages/core/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  bridge: string;
+  nativeTokenSymbol: string;
+  nativeTokenFullName: string;
+  nativeTokenDecimals: bigint;
+  startBlock: bigint;
+  blockConfirmations: number;
+  resources: Array<Resource>;
+}
+
+export type FeeHandler = {
```
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:115]: I think it works for substrate as well, no?

File: packages/core/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+  assetID?: number;
+  assetName: string;
+  xcmMultiAssetId: XcmMultiAssetIdType;
+};
+
+export type Route = {
+  fromDomain: Domain;
+  toDomain: Domain;
+  resource: Resource;
+  feeHandler: FeeHandler;
```
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:86]: #363 ^^

File: packages/core/src/index.ts (hunk 12 above)
Comment (mpetrunic) [reply packages/core/src/index.ts:102]: This is shared sygma config which doesn't contain that.
Flow is:
- fetch shared sygma config
- fetch indexer data (iconUrl)
- combine those two into Domain object (which contains iconUrl)

As for caip id, it probably makes sense to add into shared config. @MakMuftic wdyt?

File: packages/core/src/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:159]: This method is async, nothing prevents this method from fetching all domains from network/or cache if already fetched

File: packages/core/src/index.ts (hunk 12 above)
Comment (saadjhk) [inline packages/core/src/index.ts:102]: rename to `sygmaId`

File: packages/core/src/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (saadjhk) [reply packages/core/src/index.ts:115]: @mpetrunic only used in `EthereumConfig`

File: packages/core/src/index.ts (hunk 12 above)
Comment (MakMuftic) [reply packages/core/src/index.ts:102]: I would put capid to shared configuration and iconUrl we fetch from Indexer.
I can create issue to expand shared config with this if you agree @mpetrunic @saadjhk ?

File: packages/evm/src/fungible.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+  Domain,
+  EvmResource,
+  FeeHandlerType,
+} from "@buildwithsygma/sygma-sdk-core";
+import type { SecurityModel } from "../../core/src/index.js";
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+interface TransactionRequest {
+  to: string;
+  value: bigint;
```
Comment (saadjhk) [inline packages/evm/src/fungible.ts:11]: why is this needed?
the sdk can directly return `PopulatedTransaction` object and arrays

File: packages/evm/src/fungible.ts [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  maxFee?: bigint;
+};
+
+/**
+ * Return amount of liquidity tokens on resource handler
+ * @param provider
+ * @param resource
+ */
+export function getLiquidity(
+  provider: Eip1193Provider,
```
Comment (saadjhk) [inline packages/evm/src/fungible.ts:39]: shouldn't this method be a part of `EvmFungibleAssetTransfer` class? Previously it was part of `BaseAssetTransfer` class. 

It requires `Config` in its context to be able to extract handler address and then query balance of that address

Comment (MakMuftic) [reply packages/evm/src/fungible.ts:39]: Does it make sense to pass `Config` as a param to this function?

File: packages/evm/src/fungible.ts (hunk 15 above)
Comment (MakMuftic) [reply packages/evm/src/fungible.ts:11]: Isn't this `PopulatedTransaction` ethers specific type @saadjhk? If it is, then the idea is not to return lib. specific types from the SDK. cc: @mpetrunic, correct me if I misunderstood the idea.

File: packages/core/src/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (MakMuftic) [reply packages/core/src/index.ts:115]: Yep, only on Ethereum, as there are no fee handlers on the substrate and no addresses for them. It is one fee-handling pallet with different fee logic implementations.

File: packages/core/src/index.ts (hunk 12 above)
Comment (MakMuftic) [reply packages/core/src/index.ts:102]: I will create issues for this on shared config now :saluting_face:

File: packages/evm/src/fungible.ts (hunk 16 above)
Comment (saadjhk) [reply packages/evm/src/fungible.ts:39]: @MakMuftic can you point me to its use case/usage atm? then i'd be able to come up with a refined interace

File: packages/evm/src/fungible.ts (hunk 15 above)
Comment (mpetrunic) [reply packages/evm/src/fungible.ts:11]: yup,  idea is to have ability to eventually remove ethers from the lib and replace it with something smaller/less bloated

Comment (saadjhk) [reply packages/evm/src/fungible.ts:11]: @mpetrun5 contract interfaces and package is heavily dependent on ethers as well

Comment (mpetrunic) [reply packages/evm/src/fungible.ts:11]: True, but thats internal and it's easy to replace. Both web3.js and viem doesn't require typechain for types anymore and can consume json abi directly

File: packages/core/src/index.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (MakMuftic) [reply packages/core/src/index.ts:86]: After having a discussion with @saadjhk we believe having `feeHandler` doesn't make sense here. 
This would mean that the SDK for getRoutes would need to call N*2 RPC calls (where N is a number of routes).

This doesn't mean we should skip #363, as we should remove this dependency inside the `getFee` method - where SDK would do the on-chain query for this information.

@mpetrunic wdyt?

[the remark this answers is not in the document]
Comment (saadjhk) [reply packages/core/src/index.ts:86]: @mpetrunic for each route fetching the `_domainResourceIDToFeeHandlerAddress` and then the type of `feeHandler`

do you want to show the fee type e.g percentage with each route and the percentage value as well

**Review by Lykhoyda — APPROVED** [review]

File: packages/core/src/index.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:86]: ideally yes

[the remark this answers is not in the document]
Comment (mpetrunic) [reply packages/core/src/index.ts:86]: but if you think it's gonna do too much requests or be too slow, we could extract into separate method, let's say `getRouteFeeDetails`

File: packages/core/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  destinationHash: string;
+  depositNonce: number;
+  fromDomainId: number;
+  toDomainId: number;
+};
+
+export interface BaseConfig<T> {
+  sygmaId: number;
+  chainId: number;
+  caipId: number;
```
Comment (saadjhk) [inline packages/core/src/index.ts:104]: this needs to be a string IMO

File: packages/substrate/src/fungible.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  throw new Error("Method not implemented");
+}
+
+type SubstrateAssetTransferRequest = {
+  source: string | number | Domain;
+  sourceNetworkProvider: ApiPromise;
+  destination: string | number | Domain;
+  resource: string | SubstrateResource;
+  amount: bigint;
+  destinationAddress: string;
```
Comment (Lykhoyda) [inline packages/substrate/src/fungible.ts:32]: To use the init config from "core", we still need to provide the environment similar to EVM. Did I miss something about why the environment was removed? I still need to know it for the "createDestIdMultilocationData" method

File: packages/core/src/index.ts (hunk 17 above)
Comment (MakMuftic) [reply packages/core/src/index.ts]: For sure as caipID is string by standard definition, not a number

File: packages/substrate/src/fungible.ts (hunk 18 above)
Comment (saadjhk) [reply packages/substrate/src/fungible.ts:32]: this has been discussed a few times in the office and we have decided to restore environment parameter in config class, however it should now be set as an environment variable and used from there

**Comment by saadjhk** [comment]
https://github.com/sygmaprotocol/sygma-sdk/pull/375 incorporating changes from this PR was already merged so I'm closing this one
